ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Query.hs

```diff
@@ -30,28 +44,132 @@ import           Codec.CBOR.Decoding
 import           Ouroboros.Consensus.Config.SupportsNode
 import           Ouroboros.Consensus.HeaderValidation (HasAnnTip (..),
                      headerStateBlockNo, headerStatePoint)
+import           Ouroboros.Consensus.Ledger.Basics
 import           Ouroboros.Consensus.Ledger.Extended
 import           Ouroboros.Consensus.Ledger.Query.Version
 import           Ouroboros.Consensus.Node.NetworkProtocolVersion
                      (BlockNodeToClientVersion)
 import           Ouroboros.Consensus.Node.Serialisation
-                     (SerialiseNodeToClient (..), SerialiseResult (..))
+                     (SerialiseNodeToClient (..), SerialiseResult (..),
+                     SerialiseResult' (..))
+import           Ouroboros.Consensus.Storage.LedgerDB
+import qualified Ouroboros.Consensus.Storage.LedgerDB as LedgerDB
 import           Ouroboros.Consensus.Util (ShowProxy (..), SomeSecond (..))
 import           Ouroboros.Consensus.Util.DepPair
+import           Ouroboros.Consensus.Util.IOLike
 import           Ouroboros.Network.Block (HeaderHash, Point (..), StandardHash,
                      decodePoint, encodePoint)
 import           Ouroboros.Network.Protocol.LocalStateQuery.Type
-                     (ShowQuery (..))
+
+{-------------------------------------------------------------------------------
+  Footprints
+-------------------------------------------------------------------------------}
+
+-- | Queries on the local state might require reading ledger tables from disk.
+-- This datatype (which will sometimes be concretized via @sing@) allows
+-- Consensus to categorize the queries.
+data QueryFootprint =
+    -- | The query doesn't need ledger tables, thus can be answered only with
+    -- the ledger state.
+    QFNoTables
+    -- | The query needs some tables, but doesn't need to traverse the whole
+    -- backing store.
+  | QFLookupTables
+    -- | The query needs to traverse the whole backing store.
+  | QFTraverseTables
+
+type instance Sing = SQueryFootprint
+
+type SQueryFootprint :: QueryFootprint -> Type
+data SQueryFootprint a where
+    SQFNoTables       :: SQueryFootprint QFNoTables
+    SQFLookupTables   :: SQueryFootprint QFLookupTables
+    SQFTraverseTables :: SQueryFootprint QFTraverseTables
+
+instance SingI QFNoTables where
+  sing = SQFNoTables
+instance SingI QFLookupTables where
+  sing = SQFLookupTables
+instance SingI QFTraverseTables where
+  sing = SQFTraverseTables
+
+type SomeBlockQuery :: (QueryFootprint -> Type -> Type) -> Type
+data SomeBlockQuery q =
+  forall footprint result. SingI footprint => SomeBlockQuery (q footprint result)
+
+{-------------------------------------------------------------------------------
+  Block Queries
+-------------------------------------------------------------------------------}
+
+-- | Different queries supported by the ledger, indexed by the result type.
+type BlockQuery :: Type -> QueryFootprint ->  Type -> Type
+data family BlockQuery
```

Comment:
Could you remind me why we need the footprint type tag + singletons? Could the same separation into three types of queries been achieved using three data families?

```haskell
data BlockQuery blk result =
    BlockQueryNoTables BlockQueryNoTables
  | BlockQueryLookupTables BlockQueryLookupTables
  | BlockQueryTaverseTables BlockQueryTaverseTables
  
data family BlockQueryNoTables blk result
data family BlockQueryLookupTables blk result
data family BlockQueryTaverseTables blk result
```

Reply:
Maybe, but that might have been a bigger rework. It might save us some duplication though 🤔

Reply:
#1326